mesonbuild/scripts/dist.py

```diff
@@ -0,0 +1,123 @@
+# Copyright 2017 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import os, sys
+import shutil
+import argparse
+import subprocess
+import pickle
+import hashlib
+import tarfile, zipfile
+import tempfile
+from glob import glob
+from mesonbuild.environment import detect_ninja
+
+def create_hash(fname):
+    hashname = fname + '.sha256sum'
+    m = hashlib.sha256()
+    m.update(open(fname, 'rb').read())
+    with open(hashname, 'w') as f:
+        f.write('%s %s\n' % (m.hexdigest(), os.path.split(fname)[-1]))
+
+def create_zip(zipfilename, packaging_dir):
+    prefix = os.path.split(packaging_dir)[0]
+    removelen = len(prefix)+1
+    zf = zipfile.ZipFile(zipfilename, 'w', compression=zipfile.ZIP_DEFLATED,
+                         allowZip64=True)
+    zf.write(packaging_dir, packaging_dir[removelen:])
+    for root, dirs, files in os.walk(packaging_dir):
+        for d in dirs:
+            dname = os.path.join(root, d)
+            zf.write(dname, dname[removelen:])
+        for f in files:
+            fname = os.path.join(root, f)
+            zf.write(fname, fname[removelen:])
+    zf.close()
+
+def create_dist(dist_name, src_root, bld_root, dist_sub):
+    distdir = os.path.join(dist_sub, dist_name)
+    gitdir = os.path.join(src_root, '.git')
+    if os.path.exists(distdir):
+        shutil.rmtree(distdir)
+    os.makedirs(distdir)
+    dest_gitdir = os.path.join(distdir, '.git')
+    shutil.copytree(gitdir, dest_gitdir)
+    subprocess.check_call(['git', 'reset', '--hard'], cwd=distdir)
```

Comment:
I think running any git commands that change state should be considered an anti-pattern. Especially stuff like git clean and git reset are very dangerous. It's pretty much guaranteed that sooner or later people will be bitten by this.

Another problem is that some git directories can be quite large. E.g. people might have some disconnected branches in the style of `gh-pages`, or a very long history with some binaries committed by mistake, etc. Git supports creating a new work-dir very well, and that's a much clear approach.